File: shared/src/main/java/xyz/klinker/messenger/shared/util/media/MediaMessageParserFactory.java
package xyz.klinker.messenger.shared.util.media;

import android.content.Context;

import xyz.klinker.messenger.shared.util.media.parsers.ArticleParser;
import xyz.klinker.messenger.shared.util.media.parsers.YoutubeParser;

/**
 * Create an instance that can be used to parse media from the message text
 */
public class MediaMessageParserFactory {

    public MediaParser getInstance(Context context, String messageText) {
        MediaParser[] parsers = buildParsers(context);

        for (MediaParser parser : parsers) {
            if (parser.canParse(messageText)) {
                return parser;
            }
        }

        return null;
    }

    private MediaParser[] buildParsers(Context context) {
        return new MediaParser[] {
                new YoutubeParser(context),
                new ArticleParser(context)
        };
    }
}

File: shared/src/main/java/xyz/klinker/messenger/shared/util/media/MediaMessageParserFactory.kt
package xyz.klinker.messenger.shared.util.media

import android.content.Context

import xyz.klinker.messenger.shared.util.media.parsers.ArticleParser
import xyz.klinker.messenger.shared.util.media.parsers.YoutubeParser

/**
 * Create an instance that can be used to parse media from the message text
 */
class MediaMessageParserFactory {

    fun getInstance(context: Context, messageText: String): MediaParser? {
        return buildParsers(context).firstOrNull { it.canParse(messageText) }
    }

    private fun buildParsers(context: Context): Array<MediaParser> {
        return arrayOf(YoutubeParser(context), ArticleParser(context))
    }
}
